### ISSUE TITLE
Timed Robot doesn't use commands

### ISSUE THREAD

**Issue by WarrenReynolds** [JockB/RPB-Simulator#1]
Hi Jock,
I was just having a look at your code when I realized the issue. You don't use commands if your robot project is based on the timed robot starting template.  The timed robot is only used for very basic robot code.  As soon as you want to do anything simultaneously you are much better to move to command based.

**Comment by WarrenReynolds** [comment]
Hi Jock, 
The wizard that creates the blank robot projects also has setting in it that modify the include paths for the system include files ie the one found when using #include <> as compared to the found with #include "".  So you won't find <frc2/command/Command.h> if you try and include it in your project because your project was initially created as a timed robot project, not a command based project.

I'll go over the differences between timed robot and command based robot to hopefully shine a light on why we used command based or timed robot.  

In the timed only model, all of your control logic and program flows stems from teleop periodic.  If you are just reading a joystick, updating drive motor powers, and turning a solenoid on or off, you can easily fit all that logic into teleop periodic.  This code is easy to read and understand and hence this is why a lot of the sample code is based on the timed robot project.

But as soon as you start to do anything complicated, like respond to every joystick button for all the different options your robot can do like, put the intake in and out.  Start and stop the intake roller, reverse the intake roller if the current goes to high and it jams, vary the speed of the intake roller as the robot changes speed, turn on the motor in the magazine to advance the balls, drive an arm to exact angle and spin up a shooter to an exact RPM all based of a distance from target and then feed balls into a shooter at the correct speed and follow a motion profile on the drive wheels so that the robot follows a pre-programmed path and also look at a vision target to correct the heading if the robot is off target, and the change all the joystick buttons handle the robot end game like climbing up a step or onto a bar.

As you could image to write all of this code in a non blocking format in a single sub routine (teleop periodic) using 100's of if-then-else statements would be quiet a challenge and the code would become very unmanageable very quickly, as well as un-reusable.

This is where command based shines.  Every single one of those tasks are broken up into commands that control subsystems.  The execution logic of each command is handled in the commands own initialize(), execute(), isfinished(), end() methods.  These commands are then attached to joystick buttons and other control logic.  Commands can be sequentially added together, run in parallel or a mixture of both. The only thing that the robot periodic task need to do is tell the scheduler to run. (ie go through the list of active commands and run their respective methods) All of this command processing is happening in parallel (actually not true, each command is processed individually but because of the structure of commands they appear to be doing everything in parallel).
But the down side of all this is that because the execution flow is jumping all of the place and procedures are going in and out of scope continuously, pointers to all sorts of things need to be stored in member variables of the classes that the commands and subsystems are made out of so that everything can talk to what it needs to talk to.  Trust me it does become clear after awhile and it is a huge penny drop moment when it does.
Hope this helps.  I'm finally starting to feel a little better today, so I'm hoping to be on deck come Monday at the start of the season.

**Comment by JockB** [comment]
Thank you Warren

Glad to here you are on the improve.  And thank you for all the explanation about Command v Timed.

I think I am gradually getting a better understanding and have now restarted my Simulator project as a command based project.  I thought I had some understanding of the command based process but had also noticed that a command based project #includes TimedRobot.h (at least I think it does).  

In my restarted project I thought I would add little bits and keep rebuilding it and make sure it kept succeeding after each step but in trying to correct any fault that crept in I have strayed further and further from the true path.  In addition as I have copied and pasted from a running sample every now and then I pasted to the wrong project and have now also corrupted my sample program.

What I am trying to do at the moment is set up a project that 
1. can be deployed to the test table, 
2. has the basic command format, 
3. takes input from the joystick,
4. outputs variable values, strings (and possibly other things) to the SmartDashboard, and
5. has a place I can write my simulation ideas where I can put all or part of these into initialize(), execute(), isfinished(), end() methods.

I was hoping that would be the simple bit before I started what I expected to be the hard bit of doing the simulation ideas.  So far I am lost somewhere around no. 2.

Can you tell please how to send my project to you via github so that only you can read it please?

Regards

Jock



________________________________________

**Comment by WarrenReynolds** [comment]
Yes a command based robot class is still templated from the timed robot base class because this is what looks after the 20ms loops etc.  But literally there is only one line of code that lives in the periodic methods of the various operating modes and that is to tell the scheduler to run.  The only other stuff in there is about which command to schedule when Autonomous starts and when to stop the autonomous command (ie when teleop starts) 

To publish to a repository.
Click the source control icon on the left side (usually the 3rd one down)
Enter text into the commit description (usually use Initial commint for the first commit) and then click on the tick.
Answer yes to the question about adding staged changes.

Then on the blue bar across the bottom on VS code, click the arrow facing up.
You can choose either private or public.  If you choose private you will need to add me so that I can see it.

Then go to github and copy the URL for the new repo.  Should be something like github.com/jockb/newprojectname

Send that to me and I can take a look at it.

**Comment by JockB** [comment]
Thank you Warren,

I could not see where I "will need to add" you but otherwise followed your directions.

Here is the URL:
https://github.com/JockB/RPB-Command-Simulator

The hieroglyphics means there are comments at error spots at :
MyDriveDistanceCommand.h line 20
RobotContainer.h line 39
RobotContainer.cpp line 18 and
I also noted RobotContainer.h line 35 but I cannot find that now.

Regards

Jock



________________________________________

**Comment by WarrenReynolds** [comment]
I cant access that Repo.  You might have made it a private repository when you pushed it to github.  If it is private you will need to add me as a collaborator.  My github username is WarrenReynolds.

**Comment by JockB** [comment]
I believe I have now invited you to collaborate.

________________________________________
